### PR TITLE
Refactor sold class

### PR DESCRIPTION
**Pull request by akawashiro**
- Split out Sold class into sold.h
- Write comments in some difficult (for me) functions

### CODE REVIEW

File: symtab_builder.cc [hunk 1]
```diff
@@ -42,6 +44,13 @@ uintptr_t SymtabBuilder::AddSym(const Syminfo& sym) {
     return index;
 }
 
+// Returns true when the symbol specified with (name, soname, version) is
+// defined.
+// TODO(akawashiro) Is it true? Is there any defined symbols whose st_value is
```
Comment (shinh) [inline symtab_builder.cc:49]: I think the intention of this function is what you wrote. However, the implementation looks wrong to me (sorry!). As you wrote in this TODO, I think there are some defined symbols with `st_value==0`. `IsDefined` corresponds to https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/elf/dl-lookup.c#L78 and symbols can be defined even if `st_value` is zero when `sym->st_shndx != SHN_ABS` or `stt != STT_TLS`.

I think https://github.com/shinh/sold/blob/83fdbfa24c34c7cf3a22d7f96bdcd0688c8b0fda/symtab_builder.cc#L78 should be corrected to `IsDefined(sym.sym)`, but I'm not 100% sure

File: symtab_builder.cc [hunk 2]
```diff
@@ -42,6 +44,13 @@ uintptr_t SymtabBuilder::AddSym(const Syminfo& sym) {
     return index;
 }
 
+// Returns true when the symbol specified with (name, soname, version) is
```
Comment (shinh) [inline symtab_builder.cc:47]: How about

Returns true and fills `st_value` to `value_or_index`

?

File: symtab_builder.cc [hunk 3]
```diff
@@ -42,6 +44,13 @@ uintptr_t SymtabBuilder::AddSym(const Syminfo& sym) {
...
+// defined.
+// TODO(akawashiro) Is it true? Is there any defined symbols whose st_value is
+// 0?
+// When the specified symbol is not defined, SymtabBuilder::Resolve pushes it
+// to sym_ and exposed_syms_.
```
Comment (shinh) [inline symtab_builder.cc:52]: How about

and fills the index of the added symbol to `val_or_index`

?

File: sold.cc [hunk 4]
```diff
@@ -1,985 +1,623 @@
...
+    switch (type) {
+        // TODO(akawashiro) Check sym is invalid.
+        // This relocation type doesn't need any symbol therefore sym
+        // should be invalid, isn't it?
+        case R_X86_64_RELATIVE: {
+            newrel.r_addend += offset;
+            break;
+        }
+
+        // Even if we found the defined symbol in src_syms_, we cannot
```
Comment (shinh) [inline sold.cc:455]: I like this comment, thanks! I think this applies to all relocation types. How about moving this to before `switch (type)`? I'd also slightly adjust phrasing or the comment

```
    // Even if we found a defined symbol in src_syms_, we cannot
    // erase the relocation entry. The address needs to be fixed at
    // runtime by ASLR function so we set RELATIVE to these resolved symbols.
    switch (type) {
```

File: sold.cc [hunk 5]
```diff
@@ -1,985 +1,623 @@
...
-                    CHECK(tls_.bin_to_index.emplace(bin, tls_.data.size()).second);
-                    tls_.data.push_back({bin, start, size, file_offset, bss_offset});
-                    tls_.memsz += phdr->p_memsz;
-                    tls_.filesz += size;
-                    bss_offset += phdr->p_memsz - size;
-                }
+// Push symbols of bin to symtab.
+// When the same symbol is already in symtab, LoadDynSymtab selects a more
+// concretely defined one.
+// TODO(akawashiro) Raise an error when two same symbols are defined.
```
Comment (shinh) [inline sold.cc:363]: I don't think this is an error. For example, it's legitimate to define `malloc` in the main binary even when it is linked against libc. Adding `LOG(INFO)` would help us debugging the logic, though.

File: sold.cc [hunk 6]
```diff
@@ -1,985 +1,623 @@
...
+        off = RemapTLS("reloc", bin, off);
+        newrel.r_offset = off + tls_offset_;
+    } else {
+        newrel.r_offset += offset;
+    }
+
+    LOG(INFO) << "Relocate " << bin->Str(sym->st_name) << " at " << rel->r_offset;
+
+    switch (type) {
+        // TODO(akawashiro) Check sym is invalid.
```
Comment (shinh) [inline sold.cc:447]: I think we cannot ask all linkers to fill invalid symbol for RELATIVE relocations? Adding `LOG(WARNING)` or something would make sense, though.

**Review by shinh — APPROVED** [review]
Thanks!!!

File: symtab_builder.cc (hunk 2 above)
Comment (akawashiro) [reply symtab_builder.cc]: I see. It looks good.

File: symtab_builder.cc (hunk 3 above)
Comment (akawashiro) [reply symtab_builder.cc]: I see.

File: sold.cc (hunk 4 above)
Comment (akawashiro) [reply sold.cc]: I see.

File: symtab_builder.cc (hunk 1 above)
Comment (akawashiro) [reply symtab_builder.cc]: OK. I fix it using `IsDefined`.

File: sold.cc (hunk 5 above)
Comment (akawashiro) [reply sold.cc]: I understood. I changed to use `LOG(INFO)`.

File: sold.h [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    void EmitShdr(FILE* fp) {
+        CHECK(ftell(fp) == ShdrOffset());
+        shdr_.EmitShdrs(fp);
+    }
+
+    void DecideOffsets();
+
+    void CollectArrays();
+
+    // Collect most concretely defined symbols from all link_binaries_ and set
```
[the remark this answers is not in the document]
Comment (akawashiro) [reply sold.h]: Oh, I see. I rewrote this comment.

File: sold.cc (hunk 6 above)
Comment (akawashiro) [reply sold.cc]: I see. I use `LOG(WARNING)` here.

**Comment by akawashiro** [comment]
Thank you for your reviews.

**Pull request merged**